File: programcreek/python/01/01-06-regular-expression-matching.py
def isMatch(s, p):

    if len(p) == 0:
        return len(s) == 0

    if len(p) == 1:
        if len(s) == 0:
            return False

        if p[0] != s[0] and p[0] != '.':
            return False

        return isMatch(s[1:], p[1:])

    if p[1] != '*':
        if len(s) == 0:
            return False

        if s[0] != p[0] and p[0] != '.':
            return False

        return isMatch(s[1:], p[1:])

    else:

        if isMatch(s, p[2:]):
            return True

        i = 0
        while i < len(s) and (s[i] == p[0] or p[0] == '.'):

            if isMatch(s[i + 1:], p[2:]):
                return True
            i += 1

        return False


print isMatch("aa","a")
print isMatch("aa","aa")
print isMatch("aaa","aa")
print isMatch("aa", "a*")
print isMatch("aa", ".*")
print isMatch("ab", ".*")
print isMatch("aab", "c*a*b")

File: programcreek/js/01/01-06-regular-expression-matching.js
function isMatch(s, p) {

    if (p.length == 0) return s.length == 0;

    if (p.length == 1 ) {
        if (s.length == 0) return false;

        if (s.charAt(0) != p.charAt(0) && p.charAt(0) != '.' ) {
            return false;
        }

        return isMatch(s.substr(1), p.substr(1));
    }

    if (p.charAt(1) != '*') {
        if (s.length == 0) return false;

        if (s.charAt(0) != p.charAt(0) && p.charAt(0) != '.' ) {
            return false;
        }

        return isMatch(s.substr(1), p.substr(1));


    }
    else {
        if (isMatch(s, p.substr(2))) return true;

        var i = 0;

        while (s.charAt(i) == p.charAt(0) || p.charAt(0) == '.') {
            if (isMatch(s.substr(i + 1), p.substr(2))) return true;
            i ++;
        }

        return false;
    }

}


console.log( isMatch("aa","a") )
console.log( isMatch("aa","aa") )
console.log( isMatch("aaa","aa") )
console.log( isMatch("aa", "a*") )
console.log( isMatch("aa", ".*") )
console.log( isMatch("ab", ".*") )
console.log( isMatch("aab", "c*a*b") )